File: js/problems/70-climbing-stairs.js
/**
 * @param {number} n
 * @return {number}
 */
const climbStairs = function (n) {
  // Method 1

  // let arr = [1, 2];

  // for (let i = 2; i < n; i++) {
  //   arr[i] = arr[i-1] + arr[i-2];
  // }

  // return arr[n-1];

  // if (n < 2) return 1;

  // Method 2
  let prev = 0;
  let cur = 1;
  let sum = 0;

  for (let i = 0; i < n; i++) {
    sum = prev + cur;
    prev = cur;
    cur = sum;
  }

  return cur;
};

const ans = climbStairs(3);

console.log(ans); // 3

File: cpp/problems/70-climbing-stairs.cpp
#include <iostream>
#include <vector>

using namespace std;
class Solution {
 public:
  int climbStairs(int n) {
    if (n == 1) return 1;

    int cur = 1;
    int prev = 0;
    int sum = 0;

    for (int i = 0; i < n; i++) {
      sum = cur + prev;
      prev = cur;
      cur = sum;
    }

    return cur;
  }
};

// dynamic programming
class Solution {
 public:
  int climbStairs(int n) {
    vector<int> dp(n + 1);

    dp[0] = 1;
    dp[1] = 1;

    if (n <= 1) return dp[n];

    for (int i = 2; i <= n; i++) {
      dp[i] = dp[i - 2] + dp[i - 1];
    }

    return dp[n];
  }
};

int main() {
  Solution solution;

  solution.climbStairs(3);

  return 0;
}